app/main.go

```diff
@@ -239,36 +264,129 @@ func (app *SpacemeshApp) cleanup(cmd *cobra.Command, args []string) (err error)
+
+	//todo: should we add all components to a single struct?
+	lg := log.New("shmekel_"+instanceName, "", "")
+
+	trtl := consensus.NewAlgorithm(50, 100)
+
+	db, err := database.NewLDBDatabase(dbStorepath, 0, 0)
+	if err != nil {
+		return err
+	}
+	st, err := state.New(common.Hash{}, state.NewDatabase(db)) //todo: we probably should load DB with latest hash
+	if err != nil {
+		return err
+	}
+	rng := rand.New(mt19937.New())
+	processor := state.NewTransactionProcessor(rng, st, lg)
+
+	mesh := mesh.NewMesh(db, db, db, &trtl, processor, lg) //todo: what to do with the logger?
+	trtl.RegisterLayerCallback(mesh.LayerCompleteCallback)
+	coinToss := consensus.WeakCoin{}
+	clock := timesync.NewTicker(timesync.RealClock{}, 5*time.Second, time.Now())
+
+	blockListener := sync.NewBlockListener(swarm, blockOracle, mesh, 1*time.Second, 1, clock, lg)
+
+	ha := hare.New(hareConfig.DefaultConfig(), swarm, sgn, mesh, hareOracle, clock.Subscribe())
+
+	blockProducer := miner.NewBlockBuilder(instanceName, swarm, clock.Subscribe(), coinToss, mesh, ha, blockOracle, lg)
+
+	app.blockProducer = &blockProducer
+	app.blockListener = blockListener
+	app.mesh = mesh
+	app.clock = clock
+	app.state = st
+	app.db = db
+	app.hare = ha
+	app.P2P = swarm
+
+	return nil
+}
+
+func (app *SpacemeshApp) startServices() {
+	app.blockListener.Start()
+	err := app.hare.Start()
+	if err != nil {
+		panic("cannot start hare")
+	}
+	err = app.blockProducer.Start()
+	if err != nil {
+		panic("cannot start block producer")
+	}
+	app.clock.Start()
+}
+
+func (app *SpacemeshApp) stopServices() {
+	if app != nil {
+
+	}
+	app.clock.Stop()
+	err := app.blockProducer.Stop()
+	if err != nil {
+		log.Error("cannot stop block producer %v", err)
+	}
+	app.hare.Close() //todo: need to add this
+	app.blockListener.Close()
+
+	app.db.Close()
+}
+
 func (app *SpacemeshApp) startSpacemesh(cmd *cobra.Command, args []string) {
 	log.Info("Starting Spacemesh")
 
 	// start p2p services
 	log.Info("Initializing P2P services")
 	swarm, err := p2p.New(Ctx, app.Config.P2P)
-
 	if err != nil {
 		log.Error("Error starting p2p services, err: %v", err)
 		panic("Error starting p2p services")
 	}
 
-	app.P2P = swarm
+	// todo : register all protocols
+
+	sgn := hare.NewMockSigning() //todo: shouldn't be any mock code here
+	pub, _ := crypto.NewPublicKey(sgn.Verifier().Bytes())
+	bo := oracle.NewBlockOracle(1, int(app.Config.CONSENSUS.NodesPerLayer), pub.String())
+	hareOracle := oracle.NewHareOracle(1, pub.String())
+
+	apiConf := &app.Config.API
+
+	err = app.initServices("x", swarm, "/tmp/", sgn, bo, hareOracle)
 
 	if app.Config.TestMode {
 		app.setupTestFeatures()
 	}
 
-	// todo : register all protocols
+	if err != nil {
+		panic("got error starting services : " + err.Error())
+	}
+
+	app.startServices()
 
 	err = app.P2P.Start()
```

Comment:
coul'd go in startServices

Reply:
can't, it is created from outside services and passed as parameter so i'd rather the code that does that will start